src/main/java/com/javampire/openscad/color/OpenSCADColorProvider.java

```diff
@@ -2,31 +2,55 @@
 
 import com.intellij.openapi.editor.ElementColorProvider;
 import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFileFactory;
 import com.intellij.psi.util.PsiTreeUtil;
 import com.intellij.xml.util.ColorMap;
+import com.javampire.openscad.OpenSCADFileType;
 import com.javampire.openscad.psi.OpenSCADTypes;
-import com.javampire.openscad.psi.impl.OpenSCADBuiltinOpImpl;
+import com.javampire.openscad.psi.impl.OpenSCADBuiltinObjImpl;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import java.awt.*;
+import java.awt.Color;
 
 class OpenSCADColorProvider implements ElementColorProvider {
 
     @Nullable
     @Override
     public Color getColorFrom(@NotNull PsiElement element) {
-        if (element.getNode().getElementType() == OpenSCADTypes.STRING_LITERAL) {
-            final PsiElement colorParent = PsiTreeUtil.getParentOfType(element, OpenSCADBuiltinOpImpl.class);
-            if (colorParent != null && colorParent.getText().startsWith("color")) {
-                final String colorDef = element.getText();
-                return ColorMap.getColor(colorDef.substring(1, colorDef.length() - 1));
-            }
+        final PsiElement fullColorBlock = getOpenSCADBuiltinObj(element);
+        if (fullColorBlock != null) {
+            final String colorDef = element.getText();
+            return ColorMap.getColor(colorDef.substring(1, colorDef.length() - 1));
         }
+
         return null;
     }
 
     @Override
     public void setColorTo(@NotNull PsiElement element, @NotNull Color color) {
+        final PsiElement fullColorBlock = getOpenSCADBuiltinObj(element);
+        if (fullColorBlock != null) {
+            final String newText = createColorHex(color) + fullColorBlock.getChildren()[1].getText();
+
+            final PsiElement newElement = PsiFileFactory.getInstance(element.getProject())
+                    .createFileFromText("tmp.scad", OpenSCADFileType.INSTANCE, newText).getFirstChild();
+            fullColorBlock.replace(newElement);
```

Comment:
It seems overkill. I guess a better solution is to only update the argument of the color method, not the entire color invocation and the following methods. I did not try it but I am afraid this won't work well in every possible case, I fear we will loose code there.
I guess updating the ARG_ASSIGNMENT alone would be safer. Doing that, be aware of the different cases that we support and don't support I have added in the unit test https://github.com/ncsaba/idea-openscad/blob/master/src/test/testData/openscad/colorProvider.scad#L4-L9. The unit test might not be exhaustive, please add other color invocation format you are using.

Reply:
I tried to create file only with string. It's does not work because `"#FFFFFF` not a correct code for scad format. The created file contains PsiError element, and the replace method copy PsiError into origin file.

Thx for idea about unit tests.
I will invistigate and fix the bug on next week by my self. 

Thx for your responses.